torch/distributed/fsdp/flat_param.py

```diff
@@ -1398,7 +1409,16 @@ def _use_unsharded_grad_views(self) -> None:
                 f"{self.flat_param._fqns[i]} is missing",
             )
             param = getattr(module, param_name)
-            param.grad = view
+            if param.shape != view.shape:
```

Comment:
I assume this should only happen under `no_sync` mode. Do we need to assert that?

Reply:
Yes, ideally we can assert that, but at the moment, we do not propagate the `state._sync_gradients` attribute to the `FlatParamHandle`. When I wrote this PR a few days ago, I was not confident whether we should do that.

I like the design where the `state` _orchestrates_ the `FlatParamHandle`s. It becomes tricky when we have to maintain coherence between the `state` and `handles`. Currently, we do this for some things like mixed precision, and it causes some problems (as evidenced by my other PR stack).